Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+            
+            self.upgradeState = .awaitingConfirmationResponse
+            
+            let updatedHeader = self.addUpgradeHeaders(to: requestHead)
+            let completedHeader = self.addConnectionHeaders(to: updatedHeader)
+            return self.wrapOutboundOut(.head(completedHeader))
+        }
+        
+        return data
+    }
+    
+    private func addUpgradeHeaders(to requestHead: HTTPRequestHead) -> HTTPRequestHead {
+        
+        var head = requestHead
+        
+        head.headers.add(name: "Connection", value: "upgrade")
+        
+        let upgraderList = self.upgraders.map({ $0.supportedProtocol.lowercased() }).joined(separator:",")
+        head.headers.add(name: "Upgrade", value: upgraderList)
+
+        // Allow each upgrader the chance to add custom headers.
+        for upgrader in self.upgraders {
+            upgrader.addCustom(upgradeRequestHeaders: &head.headers)
+        }
+        
+        return head
+    }
+    
+    private func addConnectionHeaders(to requestHead: HTTPRequestHead) -> HTTPRequestHead {
+
+        var head = requestHead
+
+        for upgrader in self.upgraders {
+            for requiredHeader in upgrader.requiredUpgradeHeaders {
+                head.headers.add(name: "Connection", value: requiredHeader)
+            }
+        }
+        
+        return head
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        
+        guard !self.seenFirstResponse else {
+            // We're waiting for upgrade to complete: buffer this data.
+            self.receivedMessages.append(data)
+            return
+        }
+        
+        let responsePart = unwrapInboundIn(data)
+        
+        switch self.upgradeState {
+        case .awaitingConfirmationResponse:
+            self.firstResponseHeadReceived(context: context, responsePart: responsePart)
+        case .awaitingUpgrader, .upgrading, .upgradingAddingHandlers:
+            if case .end = responsePart {
+                // This is the end of the first response. Swallow it, we're buffering the rest.
+                self.seenFirstResponse = true
+                
+            }
+        case .upgraderReady(let upgrade):
+            if case .end = responsePart {
+                // This is the end of the first response, and we can upgrade. Time to kick it off.
+                self.seenFirstResponse = true
+                upgrade()
+            }
+        case .upgradeFailed:
+            // We were reentrantly called while delivering the response head. We can just pass this through.
+            context.fireChannelRead(data)
+        case .upgradeComplete:
+            preconditionFailure("Upgrade has completed but we have not seen a whole response and still got reentrantly called.")
+        case .requestRequired:
+            preconditionFailure("We are receiving an upgrade response and we have not requested the upgrade.")
```

Comment:
Ideally we'd turn all these `preconditionFailures` (including the ones in `write`) into errors instead, to avoid crashing in these cases.

Reply:
This is actually a local variable but is likely to be removed as part of other feedback.